server/src/main/java/org/opensearch/snapshots/SnapshotsService.java

```diff
@@ -267,6 +267,13 @@ public void createSnapshot(final CreateSnapshotRequest request, final ActionList
             listener.onFailure(new RepositoryException(repository.getMetadata().name(), "cannot create snapshot in a readonly repository"));
             return;
         }
+        if (repository.getMetadata() != null
+            && (repository.getMetadata().encrypted() != null || repository.getMetadata().cryptoMetadata() != null)) {
+            listener.onFailure(
+                new RepositoryException(repository.getMetadata().name(), "Snapshot creation for an encrypted repository is not supported")
```

Comment:
This seems a bit strange to me. This PR introduces client side encryption as a property of the repository, so shouldn't everything that interacts with a repository (via the repository interface) use that encryption? Looking at the next PR (#7337), the logic for using the CryptoPlugin is baked into the remote store business logic, which means that all features using the repository interface will have to do the same integration. It also looks like data is conditionally encrypted in that PR (metadata is left in plain text). That also seems strange given that encryption is a property of the repository. Shouldn't all data in an "encrypted" repository be encrypted?

Reply:
Basically, with the approach shown in #7337, I think "encryption" there is a property of the remote store feature. Then it could make sense that the remote store feature would decide which data is encrypted and which is left in plain text. If encryption is a property of the repository, then I think all data in that repository should be encrypted, and also the encryption should be transparent to all users of that repository (i.e. baked into the implementation of the repository interface).

Reply:
1. In this PR, we don't intend to enable encryption for a snapshot repository. To avoid giving an impression of an encrypted repository during a snapshot repo registration, this validation is explicitly failing the request. We will plan to put this into OS documentation as well. Enabling encryption for snapshot repos will require decent amount of changes.
2. Encryption is a heavy compute operation and has a direct impact on the overall performance. Metadata encryption was intentionally left out for this reason. In fact, I was wondering if we can skip checkpoint encryption as well. Customer data is only exposed in translog and segment files so it makes sense to secure those instead of encrypting everything. Basically, we will take a good hit on performance if we want to ensure transparency. We can probably add this in our OS documentation as well to make it known what exactly is being encrypted (secured).

Reply:
The reason why I didn't put it under remote_store block is because enabling encryption requires few key provider settings. Crypto clients are created using those settings. Now, if we move it to remote_store block, every time we create an index, we have to explicitly pass those settings. Unless we have a use case to encrypt one index on one repo and keep another index in plain text on same repo, this seems a duplicated effort. If user wants to reuse same s3 bucket for instance, we anyways have this abstraction of repository where we can create a new repo out of the same bucket.
So, now the only option left for us is to do it on a repository. Otherwise we have to expose another REST api to configure encryption and then use its identifier/name during index creation.

Reply:
Tagging @Bukhtawar @sachinpkale as well for their views on point 2.

Reply:
> Basically, we will take a good hit on performance if we want to ensure transparency.

Do you have numbers on this performance hit? Requiring all features that use the repository to hand-roll the encryption logic inside their business logic has considerable downsides IMO so this is not a design decision we should take lightly.

I also feel pretty strongly that if we put a simple `encrypted: true` property on a repository then everything written to that repository must be encrypted.

Reply:
Encryption is known to consume a lot of CPU. Numbers are present in the [design doc.](https://github.com/opensearch-project/OpenSearch/issues/6633)
Where are you proposing it to be moved then?

Reply:
>hand-roll the encryption logic inside their business logic  
What do you mean by this?

Reply:
The performance numbers I'm looking for is a comparison of a feature like remote store where in one case all files are encrypted and in the other case metadata files are written in plain text and data files are encrypted. Choosing not to encrypt metadata files is a performance optimization that adds complexity and influences the design. We should make a data-driven decision to add that optimization.

By hand-rolling the encryption logic, I mean that every feature (now and going forward) that uses a repository will have to check the encryption property of that repository, integrate with the CryptoClient, and hopefully do the right thing when it comes to encrypting and decrypting the data it reads and writes to the repository. The alternative is to build the encryption to be transparent, such that the encryption property of the repository determines the behavior of the BlobContainer interface (for example), and all features that use the repository do not need to know if client-side encryption is enabled because it works the same as the unencrypted version.

Reply:
The decision is data driven only. We have already established that encryption takes huge amount of CPU. Adding more such compute will naturally affect performance. The path we should in fact take is to minimize these computes. We have carried out multiple perf runs and scenarios where CPU is already at its peak, adding any compute had a visible impact on throughput.

Point of adding "encrypted" at repository level is exactly that. Every feature using repository must check if the repository is encrypted or not. Having said that, today we don't have encryption enabled snapshots. With this we will have a way to build or forced to build encryption on top of snapshots as well. Repository is an abstraction and remote store and snapshots are two of its stakeholders. Adding encryption context on top of repository makes sure that any new feature building on top of repository makes sure that it accounts for encryption and if it doesn't support it then it clearly calls out that it produces insecure data. The point of argument here is that whether we should encrypt entire data in specific case of remote store.
I believe that we should be highly cognizant when we think of encrypting data. There are specialized hardware instances available just to minimize the impact of encryption.
We can think of it like this that when a repository used in conjunction with remote store, encrypts only user data. Snapshot has its own way of using a repository. When a file exceeds certain capacity, it doesn't maintain a single file on repository and instead maintains smaller blocks of it. So, the only guy who can make sense of these files is snapshot. Similarly, remote store can take the decision of what should be encrypted. 

Why should encryption affect the behavior of BlobContainer? Both are two completely independent tasks. You can take a look at the second PR. There is no change in BlobContainer interface. As proposed in RFC also, I have built it as a loosely coupled module which can be tied to any type of content. It is not just limited to repository features.  We are starting to use it for repository but it can be extended to any other content type as well.

Proposal I am looking for is a place to maintain the metadata for encryption. Also, this metadata will go in cluster state so wherever you propose to put it, should have a respective cluster manager action (existing or new) responsible for maintaining it.

Reply:
I'm advocating for implementing client-side repository encryption at the repository layer so that features that use the repository do not need to changed based on whether the repository is using client-side encryption or not. I'm not suggesting that we change any interfaces or tightly couple the encryption logic to any repository implementation. I believe this could be implemented as wrapper of BlobStore and/or BlobContainer that does the encryption and decryption logic before delegating to a concrete repository implementation. The only reason I can see for _not_ doing that is so that features can exempt certain files from encryption for performance reasons. I don't believe it is sufficient to make this call by saying that encryption is CPU intensive. It matters how much this impacts performance. How much overhead does encrypting the metadata add?

Reply:
> The only reason I can see for not doing that is so that features can exempt certain files from encryption for performance reasons.

Just want to make sure my assumption here is correct. Is exempting metadata from encryption the only reason not to make encryption transparent to the users of the repository layer? PR #7337 shows all the undifferentiated work that each feature will have to do regarding wiring the CryptoClient, creating the encrypting stream, etc and it just seems to me that it would be a big win to do this once instead of per-feature.

Reply:
A good analogy to this is `dm-crypt` that encrypts the block device and all files using the device get to be encrypted as against `fs-crypt` that allows encryption at a filesystem level meaning each file can be encrypted differently and also choose not be encrypted at all.

So, now we can argue in favour of more flexibility and control, to leave the choice to the consumers of the repo, however we also need to understand if that is something that is needed now or can we build it later based on specific use cases, while keeping the contracts open. As @andrross pointed, this doesn't come for free, the changes on the integration PRs are too deep integrations and would need significant effort for consumers to integrate. Having said that if you can ensure that the integrations are light-weight and the new interfaces do most of the heavy lifting, I would still be glad considering moving on this approach. Essentially what I am looking for is moving what we call the ` CryptoClient` to still be bootstrapped in the repository module/service. Then at the `BlobContainer` provide mechanisms to encypt and controls to opt-out as needed.

Having said that, I usually would advocate for simplicity and maintainability wherever possible. I am not sure if security/compliance needs metadata to be encrypted as well. Given metadata is usually really small, I doubt that there would be a substantial CPU implication. Even if there is 1-2% moderate CPU increase, it might still be worth the encryption.

Reply:
I can run some perfs to record the difference between selective encryption and repo level encryption. Depending upon the results we can take a call.

> Then at the BlobContainer provide mechanisms to encypt and controls to opt-out as needed.

I would like to bring into notice few things so that we can make an informed decision. Repository and encryption have very different behaviors. For e.g. methods like delete, listBlobs, blobExists and similar methods have no relation with encryption. Also, read and writes cannot be used as they are and would require an offset adjustment and file header reads for ops acting on partial content. On top of that, it would require a completely new plugin to be built (encrypted repo plugin) if we push it at repository level. This is because the current layout of repository implementation is RepositoryPlugin -> Repository -> BlobStore -> BlobContainer. For adding crypto around BlobContainer, it would take a completely new delegation layer right to the top till RepositoryPlugin which will force us to build a new plugin unless we think of a different repository design. Please note that the new plugin we are currently building here has no relation with repository as of now. Check this [PR](https://github.com/opensearch-project/OpenSearch/pull/8465/files#diff-d842336bcaebd5732be721f46d6ac8fe53cfa1e75605ef75e8db81ab9e037fd1).

If we are saying that when we talk about data in OpenSearch, we are always going to associate it with repository then I too agree with the whole approach and in this case we will have to modify existing repository abstractions to accommodate partial encryption and decryption use cases. Otherwise, we will not be able to account for non-repository encryption use cases. For e.g., if tomorrow we want to provide encryption and decryption support for local indexes as well then we will need a non-repo flavor of crypto. This means we would need both encrypted repository plugin to support repository encryption and a separate crypto plugin to support local index encryption. Crypto plugin can probably be forked out for reusability and encrypted repository plugin can depend on crypto plugin. Again, I still think that we are trying to fit two incompatible pieces together where repository is a storage layer and encryption is a transformation layer.

Still, keeping an open mind to hear your thoughts around this. Nevertheless, I think PRs https://github.com/opensearch-project/OpenSearch/pull/8465 and https://github.com/opensearch-project/OpenSearch/pull/8466 would still be required irrespective of any route we take.